pkg/network/master/util/subnet_allocator.go

```diff
@@ -0,0 +1,204 @@
+	netMaskSize, addrLen := network.Mask.Size()
+	if hostBits == 0 {
+		return nil, fmt.Errorf("host capacity cannot be zero.")
+	} else if hostBits > uint32(addrLen-netMaskSize) {
+		return nil, fmt.Errorf("subnet capacity cannot be larger than number of networks available.")
+	}
+	subnetBits := uint32(addrLen-netMaskSize) - hostBits
+
+	snr := &subnetAllocatorRange{
+		network:    network,
+		hostBits:   hostBits,
+		subnetBits: subnetBits,
+		next:       0,
+		allocMap:   make(map[string]bool),
+	}
+
+	// In the simple case, the subnet part of the 32-bit IP address is just the subnet
+	// number shifted hostBits to the left. However, if hostBits isn't a multiple of
+	// 8, then it can be difficult to distinguish the subnet part and the host part
+	// visually. (Eg, given network="10.1.0.0/16" and hostBits=6, then "10.1.0.50" and
+	// "10.1.0.70" are on different networks.)
+	//
+	// To try to avoid this confusion, if the subnet extends into the next higher
+	// octet, we rotate the bits of the subnet number so that we use the subnets with
+	// all 0s in the shared octet first. So again given network="10.1.0.0/16",
+	// hostBits=6, we first allocate 10.1.0.0/26, 10.1.1.0/26, etc, through
+	// 10.1.255.0/26 (just like we would with /24s in the hostBits=8 case), and only
+	// if we use up all of those subnets do we start allocating 10.1.0.64/26,
+	// 10.1.1.64/26, etc.
+	//
+	// (For IPv6 we just don't bother worrying about this.)
+	if addrLen == 32 && hostBits%8 != 0 && ((hostBits-1)/8 != (hostBits+subnetBits-1)/8) {
+		// leftShift is used to move the subnet id left by the number of bits that
+		// the subnet part extends into the overlap octet (which is to say, the
+		// number of bits that the host part ISN'T using in that octet). leftMask
+		// masks out the bits that get shifted left out of the subnet part
+		snr.leftShift = 8 - (hostBits % 8)
+		snr.leftMask = 1<<subnetBits - 1
+		// rightShift and rightMask are used to copy the shifted-out upper bits of
+		// the subnet id back down to the lower bits
+		snr.rightShift = subnetBits - snr.leftShift
+		snr.rightMask = 1<<snr.leftShift - 1
+	}
+
+	return snr, nil
+}
+
+// markAllocatedNetwork marks network as being in use, if it is part of snr's range.
+// It returns whether the network was in snr's range.
+func (snr *subnetAllocatorRange) markAllocatedNetwork(network *net.IPNet) bool {
+	str := network.String()
+	if snr.network.Contains(network.IP) {
+		snr.allocMap[str] = true
+	}
+	return snr.allocMap[str]
+}
+
+// allocateNetwork returns a new subnet, or nil if the range is full
+func (snr *subnetAllocatorRange) allocateNetwork() *net.IPNet {
+	netMaskSize, addrLen := snr.network.Mask.Size()
+	numSubnets := uint32(1) << snr.subnetBits
+	if snr.subnetBits > 24 {
+		// If snr.subnetBits > 32, then numSubnets will have overflowed. (And if
+		// snr.subnetBits + snr.hostBits%8 >= 32 then the math below will
+		// eventually overflow.) "16M subnets ought to be enough for anybody."
```

Comment:
Hmm, I can't convince myself not to worry about "if snr.subnetBits + snr.hostBits%8 >= 32 then the math below will eventually overflow". Could you be more explicit?